File: src/services/github_service.ts
import { logTool, logServer } from '../core/logger';
import { configRepo, auditRepo } from '../db/db';

const CONFIG_KEY = 'github_auth';

export interface GitHubAuthData {
  accessToken: string;
  tokenType?: string;
  scope?: string;
  login?: string;
  name?: string;
  email?: string;
  avatarUrl?: string;
  htmlUrl?: string;
  publicRepos?: number;
  updatedAt?: number;
}

export interface GitHubRepoItem {
  id: number;
  name: string;
  fullName: string;
  description?: string;
  htmlUrl: string;
  isPrivate: boolean;
  stargazersCount: number;
  forksCount: number;
  language?: string;
  updatedAt: string;
}

export interface GitHubIssueResult {
  id: number;
  number: number;
  title: string;
  htmlUrl: string;
  state: string;
  createdAt: string;
}

export interface GitHubGistResult {
  id: string;
  htmlUrl: string;
  description?: string;
  createdAt: string;
}

export class GitHubService {
  private static instance: GitHubService;
  private authData: GitHubAuthData | null = null;

  public static getInstance(): GitHubService {
    if (!GitHubService.instance) {
      GitHubService.instance = new GitHubService();
    }
    return GitHubService.instance;
  }

  constructor() {
    this.loadPersistedAuth();
  }

  /**
   * Load saved GitHub credentials from SQLite or environment
   */
  public loadPersistedAuth(): GitHubAuthData | null {
    try {
      const persisted = configRepo.get<GitHubAuthData>(CONFIG_KEY);
      if (persisted && persisted.accessToken) {
        this.authData = persisted;
        logServer.info(`[GitHubService] Loaded persisted credentials for @${persisted.login || persisted.name || 'user'}`);
        return persisted;
      }
    } catch (err: any) {
      logTool.debug(`[GitHubService] Could not load SQLite auth: ${err.message}`);
    }

    const envToken = process.env.GITHUB_ACCESS_TOKEN || process.env.GITHUB_TOKEN;
    if (envToken) {
      this.authData = { accessToken: envToken.trim() };
      return this.authData;
    }

    return null;
  }

  /**
   * Save OAuth access token and user metadata to SQLite
   */
  public async saveAuth(auth: Partial<GitHubAuthData>): Promise<GitHubAuthData> {
    const current = this.loadPersistedAuth() || { accessToken: '' };
    const merged: GitHubAuthData = {
      ...current,
      ...auth,
      accessToken: (auth.accessToken || current.accessToken || '').trim(),
      updatedAt: Date.now(),
    };

    if (merged.accessToken) {
      try {
        const userInfo = await this.fetchApiUser(merged.accessToken);
        if (userInfo) {
          merged.login = userInfo.login || merged.login;
          merged.name = userInfo.name || merged.name;
          merged.email = userInfo.email || merged.email;
          merged.avatarUrl = userInfo.avatar_url || merged.avatarUrl;
          merged.htmlUrl = userInfo.html_url || merged.htmlUrl;
          merged.publicRepos = userInfo.public_repos ?? merged.publicRepos;
        }
      } catch (e: any) {
        logTool.debug(`[GitHubService] Auto-enrich user info notice: ${e.message}`);
      }
    }

    this.authData = merged;
    configRepo.set(CONFIG_KEY, merged);
    logServer.info(`[GitHubService] Saved GitHub credentials in SQLite (@${merged.login || merged.name || 'connected'})`);
    auditRepo.log('GITHUB', 'info', 'GitHub credentials updated', { login: merged.login, email: merged.email });
    return merged;
  }

  /**
   * Disconnect GitHub credentials
   */
  public disconnect(): void {
    this.authData = null;
    configRepo.delete(CONFIG_KEY);
    logServer.info('[GitHubService] GitHub credentials disconnected.');
    auditRepo.log('GITHUB', 'info', 'GitHub credentials disconnected');
  }

  /**
   * Get current auth status
   */
  public getStatus(): {
    connected: boolean;
    login?: string;
    name?: string;
    email?: string;
    avatarUrl?: string;
    publicRepos?: number;
  } {
    const auth = this.authData || this.loadPersistedAuth();
    return {
      connected: !!auth?.accessToken,
      login: auth?.login,
      name: auth?.name,
      email: auth?.email,
      avatarUrl: auth?.avatarUrl,
      publicRepos: auth?.publicRepos,
    };
  }

  public getAccessToken(): string {
    return this.authData?.accessToken || this.loadPersistedAuth()?.accessToken || process.env.GITHUB_ACCESS_TOKEN || process.env.GITHUB_TOKEN || '';
  }

  /**
   * Generate GitHub OAuth 2.0 Authorization URL
   */
  public getAuthorizationUrl(redirectUri?: string, clientId?: string, state?: string): string {
    const id = (clientId || process.env.GITHUB_CLIENT_ID || '').trim();
    if (!id) {
      throw new Error('GitHub Client ID is missing. Please configure GITHUB_CLIENT_ID in .env or settings.');
    }
    const stateToken = state || `jarvis_gh_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const scopes = ['repo', 'read:user', 'user:email', 'workflow', 'gist'];
    const paramObj: Record<string, string> = {
      client_id: id,
      state: stateToken,
      scope: scopes.join(' '),
      allow_signup: 'true',
    };
    const uri = redirectUri || process.env.GITHUB_REDIRECT_URI;
    if (uri && uri !== 'auto' && uri !== 'none') {
      paramObj.redirect_uri = uri;
    }
    const params = new URLSearchParams(paramObj);
    return `https://github.com/login/oauth/authorize?${params.toString()}`;
  }

  /**
   * Exchange authorization code for GitHub access token
   */
  public async exchangeAuthCode(
    code: string,
    redirectUri: string,
    clientId?: string,
    clientSecret?: string
  ): Promise<GitHubAuthData> {
    const id = (clientId || process.env.GITHUB_CLIENT_ID || '').trim();
    const secret = (clientSecret || process.env.GITHUB_CLIENT_SECRET || '').trim();

    if (!id || !secret) {
      throw new Error('GitHub Client ID and Client Secret are required for code exchange.');
    }

    const body = new URLSearchParams({
      client_id: id,
      client_secret: secret,
      code,
      redirect_uri: redirectUri,
    });

    const res = await fetch('https://github.com/login/oauth/access_token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        Accept: 'application/json',
        'User-Agent': 'JARVIS-OS/1.0',
      },
      body: body.toString(),
    });

    if (!res.ok) {
      const errText = await res.text();
      throw new Error(`GitHub token exchange failed (HTTP ${res.status}): ${errText}`);
    }

    const data = await res.json();
    if (data.error) {
      throw new Error(`GitHub token error: ${data.error_description || data.error}`);
    }

    const accessToken = data.access_token;
    if (!accessToken) {
      throw new Error('Invalid response from GitHub: missing access_token');
    }

    const saved = await this.saveAuth({
      accessToken,
      tokenType: data.token_type,
      scope: data.scope,
    });

    return saved;
  }

  private async fetchApiUser(token: string): Promise<any> {
    const res = await fetch('https://api.github.com/user', {
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: 'application/vnd.github.v3+json',
        'User-Agent': 'JARVIS-OS/1.0',
      },
    });
    if (res.ok) {
      return res.json();
    }
    throw new Error(`GitHub User API returned HTTP ${res.status}`);
  }

  /**
   * 1. Get Authenticated User Profile
   */
  public async getMyProfile(): Promise<any> {
    const token = this.getAccessToken();
    if (!token) {
      throw new Error('GitHub is not connected. Please provide an OAuth token or run github_oauth_flow.py');
    }
    return this.fetchApiUser(token);
  }

  /**
   * 2. List Authenticated User Repositories
   */
  public async listMyRepos(limit = 10, sort: 'updated' | 'created' | 'pushed' = 'updated'): Promise<GitHubRepoItem[]> {
    const token = this.getAccessToken();
    const headers: Record<string, string> = {
      Accept: 'application/vnd.github.v3+json',
      'User-Agent': 'JARVIS-OS/1.0',
    };
    if (token) headers['Authorization'] = `Bearer ${token}`;

    const res = await fetch(`https://api.github.com/user/repos?sort=${sort}&per_page=${limit}&affiliation=owner,collaborator`, { headers });
    if (!res.ok) {
      throw new Error(`Failed to list repositories (HTTP ${res.status})`);
    }

    const repos = await res.json();
    return repos.map((r: any) => ({
      id: r.id,
      name: r.name,
      fullName: r.full_name,
      description: r.description,
      htmlUrl: r.html_url,
      isPrivate: r.private,
      stargazersCount: r.stargazers_count,
      forksCount: r.forks_count,
      language: r.language,
      updatedAt: r.updated_at,
    }));
  }

  /**
   * 3. Create an Issue on a Repository
   */
  public async createIssue(owner: string, repo: string, title: string, body?: string, labels?: string[]): Promise<GitHubIssueResult> {
    const token = this.getAccessToken();
    if (!token) {
      throw new Error('GitHub write permissions require an authenticated OAuth token.');
    }

    const payload: any = { title, body: body || '' };
    if (labels && labels.length > 0) payload.labels = labels;

    const res = await fetch(`https://api.github.com/repos/${owner}/${repo}/issues`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: 'application/vnd.github.v3+json',
        'Content-Type': 'application/json',
        'User-Agent': 'JARVIS-OS/1.0',
      },
      body: JSON.stringify(payload),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`GitHub create issue failed (HTTP ${res.status}): ${err}`);
    }

    const data = await res.json();
    logTool.info(`[GitHubService] Created issue #${data.number} on ${owner}/${repo}`);
    auditRepo.log('GITHUB', 'info', `Issue #${data.number} created on ${owner}/${repo}`, { url: data.html_url });

    return {
      id: data.id,
      number: data.number,
      title: data.title,
      htmlUrl: data.html_url,
      state: data.state,
      createdAt: data.created_at,
    };
  }

  /**
   * 4. Create a Gist
   */
  public async createGist(description: string, filename: string, content: string, isPublic = false): Promise<GitHubGistResult> {
    const token = this.getAccessToken();
    if (!token) {
      throw new Error('Gist creation requires GitHub OAuth credentials.');
    }

    const payload = {
      description,
      public: isPublic,
      files: {
        [filename]: { content },
      },
    };

    const res = await fetch('https://api.github.com/gists', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: 'application/vnd.github.v3+json',
        'Content-Type': 'application/json',
        'User-Agent': 'JARVIS-OS/1.0',
      },
      body: JSON.stringify(payload),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`Failed to create Gist (HTTP ${res.status}): ${err}`);
    }

    const data = await res.json();
    logTool.info(`[GitHubService] Created gist: ${data.html_url}`);
    auditRepo.log('GITHUB', 'info', 'GitHub gist created', { url: data.html_url });

    return {
      id: data.id,
      htmlUrl: data.html_url,
      description: data.description,
      createdAt: data.created_at,
    };
  }

  /**
   * 5. Get Repository Details (with resilient public fallback)
   */
  public async getRepoDetails(owner: string, repo: string): Promise<any> {
    const token = this.getAccessToken();
    const headers: Record<string, string> = {
      Accept: 'application/vnd.github.v3+json',
      'User-Agent': 'JARVIS-OS/1.0',
    };
    if (token) headers['Authorization'] = `Bearer ${token}`;

    let res = await fetch(`https://api.github.com/repos/${owner}/${repo}`, { headers });
    if (!res.ok && res.status === 401 && token) {
      // Fallback without invalid auth header for public repositories
      res = await fetch(`https://api.github.com/repos/${owner}/${repo}`, {
        headers: { Accept: 'application/vnd.github.v3+json', 'User-Agent': 'JARVIS-OS/1.0' },
      });
    }

    if (!res.ok) {
      throw new Error(`Failed to fetch repo ${owner}/${repo} (HTTP ${res.status})`);
    }
    return res.json();
  }

  /**
   * 6. Search Repositories (Authenticated with 5000 req/hr rate limit + fallback)
   */
  public async searchRepositories(query: string, limit = 5): Promise<GitHubRepoItem[]> {
    const token = this.getAccessToken();
    const headers: Record<string, string> = {
      Accept: 'application/vnd.github.v3+json',
      'User-Agent': 'JARVIS-OS/1.0',
    };
    if (token) headers['Authorization'] = `Bearer ${token}`;

    let res = await fetch(`https://api.github.com/search/repositories?q=${encodeURIComponent(query)}&sort=stars&per_page=${limit}`, { headers });
    if (!res.ok && res.status === 401 && token) {
      res = await fetch(`https://api.github.com/search/repositories?q=${encodeURIComponent(query)}&sort=stars&per_page=${limit}`, {
        headers: { Accept: 'application/vnd.github.v3+json', 'User-Agent': 'JARVIS-OS/1.0' },
      });
    }

    if (!res.ok) {
      throw new Error(`GitHub search failed (HTTP ${res.status})`);
    }

    const data = await res.json();
    return (data.items || []).map((r: any) => ({
      id: r.id,
      name: r.name,
      fullName: r.full_name,
      description: r.description,
      htmlUrl: r.html_url,
      isPrivate: r.private,
      stargazersCount: r.stargazers_count,
      forksCount: r.forks_count,
      language: r.language,
      updatedAt: r.updated_at,
    }));
  }
}

export const githubService = GitHubService.getInstance();

File: core_engine/github_service.py
"""
GitHub Integration & OAuth 2.0 Service for J.A.R.V.I.S. Python Core Engine.
Provides autonomous repository management, issue creation, Gist sharing,
profile synchronization, and OAuth 2.0 token handling.
"""

import os
import json
import time
import sqlite3
import httpx
from pathlib import Path
from typing import Dict, Any, Optional, List
from urllib.parse import urlencode

DATA_DIR = os.path.join(os.getcwd(), "data")
TOKEN_FILE_PATH = os.path.join(DATA_DIR, "github_auth.json")
DB_PATH = os.path.join(DATA_DIR, "jarvis.db")
CONFIG_KEY = "github_auth"


class GitHubService:
    _instance = None

    @classmethod
    def get_instance(cls) -> "GitHubService":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def __init__(self):
        self.auth_data: Optional[Dict[str, Any]] = None
        self.load_persisted_auth()

    def is_valid_token(self, token: Optional[str]) -> bool:
        if not token or not isinstance(token, str):
            return False
        cleaned = token.strip()
        if not cleaned:
            return False
        # Discard mock/placeholder test tokens
        if cleaned.startswith("gho_mock_") or cleaned.startswith("test_"):
            return False
        return True

    def load_persisted_auth(self) -> Optional[Dict[str, Any]]:
        """Load GitHub credentials from JSON file, SQLite database, or environment."""
        # 1. JSON file
        try:
            if os.path.exists(TOKEN_FILE_PATH):
                with open(TOKEN_FILE_PATH, "r", encoding="utf-8") as f:
                    data = json.load(f)
                token = (data.get("accessToken") or data.get("access_token", "")).strip()
                if self.is_valid_token(token):
                    self.auth_data = {
                        "accessToken": token,
                        "tokenType": data.get("tokenType") or data.get("token_type", "bearer"),
                        "scope": data.get("scope", "repo read:user user:email workflow gist"),
                        "login": data.get("login", ""),
                        "name": data.get("name", ""),
                        "email": data.get("email", ""),
                        "avatarUrl": data.get("avatarUrl") or data.get("avatar_url", ""),
                        "htmlUrl": data.get("htmlUrl") or data.get("html_url", ""),
                        "publicRepos": data.get("publicRepos") or data.get("public_repos", 0),
                        "updatedAt": data.get("updatedAt", int(time.time() * 1000)),
                    }
                    os.environ["GITHUB_ACCESS_TOKEN"] = token
                    return self.auth_data
        except Exception:
            pass

        # 2. SQLite jarvis.db configs table
        if os.path.exists(DB_PATH):
            try:
                conn = sqlite3.connect(DB_PATH)
                cur = conn.cursor()
                cur.execute(
                    "SELECT key, value_json FROM configs WHERE key IN ('github_auth', 'github_oauth_credentials') ORDER BY updated_at DESC"
                )
                rows = cur.fetchall()
                conn.close()
                for key, val in rows:
                    if val:
                        try:
                            db_data = json.loads(val)
                            token = (db_data.get("accessToken") or db_data.get("access_token", "")).strip()
                            if self.is_valid_token(token):
                                self.auth_data = db_data
                                os.environ["GITHUB_ACCESS_TOKEN"] = token
                                # Sync to JSON file
                                try:
                                    with open(TOKEN_FILE_PATH, "w", encoding="utf-8") as jf:
                                        json.dump(self.auth_data, jf, indent=2)
                                except Exception:
                                    pass
                                return self.auth_data
                        except Exception:
                            continue
            except Exception:
                pass

        # 3. Environment variable fallback
        env_token = (os.environ.get("GITHUB_ACCESS_TOKEN") or os.environ.get("GITHUB_TOKEN") or "").strip()
        if self.is_valid_token(env_token):
            self.auth_data = {
                "accessToken": env_token,
                "tokenType": "bearer",
                "updatedAt": int(time.time() * 1000),
            }
            return self.auth_data

        return None

    def get_access_token(self) -> str:
        if self.auth_data and self.auth_data.get("accessToken"):
            return self.auth_data["accessToken"]
        auth = self.load_persisted_auth()
        return auth.get("accessToken", "") if auth else ""

    def get_status(self) -> Dict[str, Any]:
        """Return structured connection status for frontend and agent tools."""
        auth = self.auth_data or self.load_persisted_auth()
        is_connected = bool(auth and auth.get("accessToken"))
        return {
            "connected": is_connected,
            "login": auth.get("login", "") if auth else "",
            "name": auth.get("name", "") if auth else "",
            "email": auth.get("email", "") if auth else "",
            "avatarUrl": auth.get("avatarUrl", "") if auth else "",
            "publicRepos": auth.get("publicRepos", 0) if auth else 0,
            "updatedAt": auth.get("updatedAt") if auth else None,
        }

    async def fetch_user_profile(self, token: str) -> Optional[Dict[str, Any]]:
        """Fetch user profile and primary email from GitHub API."""
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                headers = {
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github.v3+json",
                    "User-Agent": "JARVIS-Python-Core/1.0",
                }
                res = await client.get("https://api.github.com/user", headers=headers)
                if res.status_code == 200:
                    data = res.json()
                    # If email is null/private, attempt to fetch from /user/emails
                    if not data.get("email"):
                        try:
                            em_res = await client.get("https://api.github.com/user/emails", headers=headers)
                            if em_res.status_code == 200:
                                emails = em_res.json()
                                for em in emails:
                                    if em.get("primary"):
                                        data["email"] = em.get("email")
                                        break
                        except Exception:
                            pass
                    return data
        except Exception as e:
            print(f"[GitHubService] User profile fetch notice: {e}")
        return None

    async def save_auth(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Save GitHub credentials to JSON file, SQLite, and environment."""
        os.makedirs(DATA_DIR, exist_ok=True)
        current = self.load_persisted_auth() or {"accessToken": "", "updatedAt": int(time.time() * 1000)}

        token = (data.get("accessToken") or data.get("access_token") or current.get("accessToken", "")).strip()
        merged = {
            **current,
            **data,
            "accessToken": token,
            "updatedAt": int(time.time() * 1000),
        }

        if token:
            os.environ["GITHUB_ACCESS_TOKEN"] = token
            # Fetch profile to enrich metadata
            profile = await self.fetch_user_profile(token)
            if profile:
                merged["login"] = profile.get("login") or merged.get("login", "")
                merged["name"] = profile.get("name") or merged.get("name", "")
                merged["email"] = profile.get("email") or merged.get("email", "")
                merged["avatarUrl"] = profile.get("avatar_url") or merged.get("avatarUrl", "")
                merged["htmlUrl"] = profile.get("html_url") or merged.get("htmlUrl", "")
                merged["publicRepos"] = profile.get("public_repos", merged.get("publicRepos", 0))

        self.auth_data = merged

        # 1. Save JSON backup
        try:
            with open(TOKEN_FILE_PATH, "w", encoding="utf-8") as f:
                json.dump(merged, f, indent=2)
        except Exception as e:
            print(f"[GitHubService] Error saving JSON: {e}")

        # 2. Save to SQLite jarvis.db
        try:
            conn = sqlite3.connect(DB_PATH)
            cur = conn.cursor()
            cur.execute("""
                CREATE TABLE IF NOT EXISTS configs (
                    key TEXT PRIMARY KEY,
                    value_json TEXT NOT NULL,
                    updated_at INTEGER NOT NULL
                )
            """)
            cur.execute("""
                INSERT INTO configs (key, value_json, updated_at)
                VALUES (?, ?, ?)
                ON CONFLICT(key) DO UPDATE SET
                    value_json = excluded.value_json,
                    updated_at = excluded.updated_at
            """, (CONFIG_KEY, json.dumps(merged), int(time.time() * 1000)))
            conn.commit()
            conn.close()
        except Exception as e:
            print(f"[GitHubService] SQLite save warning: {e}")

        return merged

    def disconnect(self) -> None:
        """Disconnect GitHub credentials."""
        self.auth_data = None
        os.environ.pop("GITHUB_ACCESS_TOKEN", None)
        if os.path.exists(TOKEN_FILE_PATH):
            try:
                os.remove(TOKEN_FILE_PATH)
            except Exception:
                pass
        if os.path.exists(DB_PATH):
            try:
                conn = sqlite3.connect(DB_PATH)
                cur = conn.cursor()
                cur.execute("DELETE FROM configs WHERE key = ?", (CONFIG_KEY,))
                conn.commit()
                conn.close()
            except Exception:
                pass

    def get_authorization_url(self, redirect_uri: Optional[str] = None, client_id: Optional[str] = None) -> str:
        """Generate GitHub OAuth 2.0 Authorization URL."""
        cid = (client_id or os.environ.get("GITHUB_CLIENT_ID", "")).strip()
        if not cid:
            raise ValueError("GitHub Client ID is required. Please set GITHUB_CLIENT_ID in .env or pass client_id.")

        state = f"jarvis_gh_{int(time.time() * 1000)}"
        scopes = ["repo", "read:user", "user:email", "workflow", "gist"]
        params = {
            "client_id": cid,
            "state": state,
            "scope": " ".join(scopes),
            "allow_signup": "true",
        }
        if redirect_uri:
            params["redirect_uri"] = redirect_uri

        return f"https://github.com/login/oauth/authorize?{urlencode(params)}"

    async def exchange_auth_code(
        self,
        code: str,
        redirect_uri: str,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None
    ) -> Dict[str, Any]:
        """Exchange authorization code for GitHub access token."""
        cid = (client_id or os.environ.get("GITHUB_CLIENT_ID", "")).strip()
        sec = (client_secret or os.environ.get("GITHUB_CLIENT_SECRET", "")).strip()

        if not cid or not sec:
            raise ValueError("GitHub Client ID and Client Secret are required for code exchange.")

        payload = {
            "client_id": cid,
            "client_secret": sec,
            "code": code,
            "redirect_uri": redirect_uri,
        }

        async with httpx.AsyncClient(timeout=15.0) as client:
            res = await client.post(
                "https://github.com/login/oauth/access_token",
                data=payload,
                headers={"Accept": "application/json", "User-Agent": "JARVIS-Python-Core/1.0"}
            )
            if res.status_code != 200:
                raise RuntimeError(f"GitHub token exchange failed (HTTP {res.status_code}): {res.text}")

            data = res.json()
            if data.get("error"):
                raise RuntimeError(f"GitHub token error: {data.get('error_description') or data.get('error')}")

            access_token = data.get("access_token", "")
            if not access_token:
                raise RuntimeError("Invalid response from GitHub: missing access_token")

            saved = await self.save_auth({
                "accessToken": access_token,
                "tokenType": data.get("token_type", "bearer"),
                "scope": data.get("scope", ""),
            })
            return saved

    async def list_my_repos(self, limit: int = 10, sort: str = "updated") -> List[Dict[str, Any]]:
        """List authenticated user's repositories."""
        token = self.get_access_token()
        if not token:
            raise ValueError("GitHub not connected. Please authenticate first.")

        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github.v3+json",
            "User-Agent": "JARVIS-Python-Core/1.0",
        }
        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.get(
                f"https://api.github.com/user/repos?per_page={limit}&sort={sort}",
                headers=headers
            )
            if res.status_code == 200:
                raw = res.json()
                return [
                    {
                        "id": r.get("id"),
                        "name": r.get("name"),
                        "fullName": r.get("full_name"),
                        "description": r.get("description"),
                        "htmlUrl": r.get("html_url"),
                        "isPrivate": r.get("private", False),
                        "stargazersCount": r.get("stargazers_count", 0),
                        "forksCount": r.get("forks_count", 0),
                        "language": r.get("language"),
                        "updatedAt": r.get("updated_at"),
                    }
                    for r in raw
                ]
            raise RuntimeError(f"Failed to fetch repositories (HTTP {res.status_code}): {res.text}")

    async def create_issue(self, owner: str, repo: str, title: str, body: str = "", labels: Optional[List[str]] = None) -> Dict[str, Any]:
        """Create an issue in a GitHub repository."""
        token = self.get_access_token()
        if not token:
            raise ValueError("GitHub not connected. Please authenticate first.")

        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github.v3+json",
            "User-Agent": "JARVIS-Python-Core/1.0",
        }
        payload: Dict[str, Any] = {"title": title, "body": body}
        if labels:
            payload["labels"] = labels

        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.post(f"https://api.github.com/repos/{owner}/{repo}/issues", json=payload, headers=headers)
            if res.status_code == 201:
                data = res.json()
                return {
                    "id": data.get("id"),
                    "number": data.get("number"),
                    "title": data.get("title"),
                    "htmlUrl": data.get("html_url"),
                    "state": data.get("state"),
                    "createdAt": data.get("created_at"),
                }
            raise RuntimeError(f"Failed to create issue (HTTP {res.status_code}): {res.text}")

    async def create_gist(self, description: str, files: Dict[str, Dict[str, str]], is_public: bool = False) -> Dict[str, Any]:
        """Create a GitHub Gist."""
        token = self.get_access_token()
        if not token:
            raise ValueError("GitHub not connected. Please authenticate first.")

        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github.v3+json",
            "User-Agent": "JARVIS-Python-Core/1.0",
        }
        payload = {
            "description": description,
            "public": is_public,
            "files": files,
        }

        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.post("https://api.github.com/gists", json=payload, headers=headers)
            if res.status_code == 201:
                data = res.json()
                return {
                    "id": data.get("id"),
                    "htmlUrl": data.get("html_url"),
                    "description": data.get("description"),
                    "createdAt": data.get("created_at"),
                }
            raise RuntimeError(f"Failed to create Gist (HTTP {res.status_code}): {res.text}")


github_service = GitHubService.get_instance()
